doc: translate sync.md (#1656)
juicedata/juicefs: docs/en/administration/sync.md

---
sidebar_label: Synchronization
---

# Migrate and Synchronize Data across Clouds with JuiceFS Sync

> The `sync` subcommand of JuiceFS is a full-featured data synchronization utility that can synchronize or migrate data concurrently with multiple threads between all [object storages JuiceFS supports](../reference/how_to_setup_object_storage.md). It can be used to migrate data not only between _object storage_ and _JuiceFS_, but also between _object storages_ in different clouds or regions. In addition, it can also be used to synchronize local directories, access remote directories through SSH, HDFS, WebDAV, etc., quite similar to `rsync`. It also provides advanced features such as full synchronization, incremental synchronization, and conditional pattern matching.

Reviewer: ```suggestion
The `sync` subcommand of JuiceFS is a full-featured data synchronization utility that can synchronize or migrate data concurrently with multiple threads between all [object storages JuiceFS supports](../reference/how_to_setup_object_storage.md). It can be used to migrate data not only between _object storage_ and _JuiceFS_, but also between _object storages_ in different clouds or regions. In addition, similar to `rsync`, the JuiceFS `sync` subcommand can also be used to synchronize local directories and access remote directories through SSH, HDFS, WebDAV, etc.. It also provides advanced features such as full synchronization, incremental synchronization, and conditional pattern matching.
```
原本已经翻译的很好啦，其实不改动也是可以的 :) 
不过个人感觉句式有点点单一，所以我稍微改了一下某个句子的结构，并且把段中的一个"it" 换成了 the `sync` subcommand，重新强调一下我们的主语 :P  你看看有没有必要哈

## Basic Usage

### Command Syntax

```shell
juicefs sync [command options] SRC DST
```

Synchronize data from `SRC` to `DST`, capable for both directories and files.

Arguments:

- `SRC` is the source data address or path;
- `DST` is the destination address or path;
- `[command options]` are synchronization options. See [command reference](../reference/command_reference.md#juicefs-sync) for more details.

> Address syntax is `[NAME://][ACCESS_KEY:SECRET_KEY@]BUCKET[.ENDPOINT][/PREFIX]`.

Reviewer: ```suggestion
Address syntax follows `[NAME://][ACCESS_KEY:SECRET_KEY@]BUCKET[.ENDPOINT][/PREFIX]`.
```

Explanation:

- `NAME` is the storage type like `s3` or `oss`. See [available storage services](../reference/how_to_setup_object_storage.md#supported-object-storage) for more details;
- `ACCESS_KEY` and `SECRET_KEY` are the credentials for accessing object storage APIs;
- `BUCKET[.ENDPOINT]` is the address of the object storage;
- `PREFIX` is the common prefix of the directories to synchronize, optional.

> Here is a sample object storage address of Amazon S3.

Reviewer: ```suggestion
Here is an example of the object storage address of Amazon S3.
```
感觉example会更好一些？

```
s3://ABCDEFG:HIJKLMN@myjfs.s3.us-west-1.amazonaws.com
```

In particular, `SRC` and `DST` ending with a trailing `/` are treated as directories, e.g. `movie/`. Those don't end with a trailing `/` are treated as _prefixes_, and will be used for pattern matching. For example, assuming we have `test` and `text` directories in the current directory, the following command can synchronize them into the destination `~/mnt/`.

```shell
juicefs sync ./te ~/mnt/te
```

> In this way, the subcommand `sync` takes `te` as a prefix to find all the matching directories, i.e. `test` and `text`. Meanwhile, `~/mnt/te` is also a prefix. All directories and files synchronized to this destination will be renamed by replacing the original prefix `te` with the new prefix `te`, actually not changed in this example. If we take another prefix, for example `ab`,

Reviewer: ```suggestion
In this way, the subcommand `sync` takes `te` as a prefix to find all the matching directories, i.e. `test` and `text`. `~/mnt/te` is also a prefix, and all directories and files synchronized to this destination will be renamed by replacing the original prefix `te` with the new prefix `te`. The changes in the names of directories and files before and after synchronization cannot be seen in the above example. However, if we take another prefix, for example, `ab`,
```
- 感觉原来的翻译好像有些不完整，我尝试着把它写全了 （但是看着好些有些啰嗦）:P
- 不知道可不可以直接用下面的例子，省略 `~/mnt/te`的例子，好像下面的一个例子就可以解释清楚了

```shell
juicefs sync ./te ~/mnt/ab
```

> the `test` directory synchronized to destination directory will be `abst`, and `text` will be `abxt`.

Reviewer: ```suggestion
the `test` directory synchronized to the destination directory will be renamed as `abst`, and `text` will be `abxt`.
```

### Required Storages

> Assume that we have these following storages.

Reviewer: ```suggestion
Assume that we have the following storages.
```

1. **Object Storage A** <span id="bucketA" />
   - Bucket name: aaa
   - Endpoint: `https://aaa.s3.us-west-1.amazonaws.com`

2. **Object Storage B** <span id="bucketB" />
   - Bucket name: bbb
   - Endpoint: `https://bbb.oss-cn-hangzhou.aliyuncs.com`
   
3. **JuiceFS File System** <span id="bucketC" />
   - Metadata Storage: `redis://10.10.0.8:6379/1`
   - Object Storage: `https://ccc-125000.cos.ap-beijing.myqcloud.com`
   
All of the storages share the same **secret key**:

- **ACCESS_KEY**: `ABCDEFG`
- **SECRET_KEY**: `HIJKLMN`

### Synchronize between Object Storage and JuiceFS

> The following command synchronizes `movie` directory on [Object Storage A](#bucketA) to [JuiceFS File System](#bucketC).

Reviewer: ```suggestion
The following command synchronizes `movies` directory on [Object Storage A](#bucketA) to [JuiceFS File System](#bucketC).
```

```shell
# mount JuiceFS
sudo juicefs mount -d redis://10.10.0.8:6379/1 /mnt/jfs

> # synchronization

Reviewer: ```suggestion
# synchronize
```
因为前面的注释是一个动作“mount JuiceFS”，这里也用动词是不是更好些？ 或者 `# execute synchronization`? (不知道这种说话专不专业 XD

juicefs sync s3://ABCDEFG:HIJKLMN@aaa.s3.us-west-1.amazonaws.com/movies/ /mnt/jfs/movies/
```

The following command synchronizes `images` directory from [JuiceFS File System](#bucketC) to [Object Storage A](#bucketA).

```shell
# mount JuiceFS
sudo juicefs mount -d redis://10.10.0.8:6379/1 /mnt/jfs
# synchronization
juicefs sync /mnt/jfs/images/ s3://ABCDEFG:HIJKLMN@aaa.s3.us-west-1.amazonaws.com/images/
```

### Synchronize between Object Storages

The following command synchronizes all of the data on [Object Storage A](#bucketA) to [Object Storage B](#bucketB).

```shell
juicefs sync s3://ABCDEFG:HIJKLMN@aaa.s3.us-west-1.amazonaws.com oss://ABCDEFG:HIJKLMN@bbb.oss-cn-hangzhou.aliyuncs.com
```

## Advanced Usage

### Incremental and Full Synchronization

> The subcommand `sync` works incrementally by default. That is, compares the differences between the source and target paths, and then synchronizes only the differences. You can add option `--update` or `-u` to keep updated the `mtime` of the synchronized directories and files.

Reviewer: ```suggestion
The subcommand `sync` works via incremental synchronization by default. That is,  only the differences will be synchronized after comparing the source and target paths. You can add option `--update` or `-u` to keep updated the `mtime` of the synchronized directories and files.
```
之前的那句好像少了一个主语，主语在这个语境里似乎很不明确，所以我尝试着换了一下句子的结构

Author: ```suggestion
The subcommand `sync` works incrementally by default, which compares the differences between the source and target paths, and then synchronizes only the differences. You can add option `--update` or `-u` to keep updated the `mtime` of the synchronized directories and files.
```

> To do a full synchronization, i.e. always synchronize no matter whether the destination files exist or not, you can add option `--force-update` or `-f`. For example, the following command fully synchronizes `movies` directory from [Object Storage A](#bucketA) to [JuiceFS File System](#bucketC).

Reviewer: ```suggestion
For full synchronization, i.e. synchronizing all the time no matter whether the destination files exist or not, you can add option `--force-update` or `-f`. For example, the following command fully synchronizes `movies` directory from [Object Storage A](#bucketA) to [JuiceFS File System](#bucketC).
```

```shell
# mount JuiceFS
sudo juicefs mount -d redis://10.10.0.8:6379/1 /mnt/jfs
# full synchronization
juicefs sync --force-update s3://ABCDEFG:HIJKLMN@aaa.s3.us-west-1.amazonaws.com/movies/ /mnt/jfs/movies/
```

### Pattern Matching

> The pattern matching function of the `sync` subcommand is similar to that of `rsync`, which allows you to exclude or include a certain classes of files by rules, and synchronize any set of files by combining multiple rules. Now we have these following rules available.

Reviewer: ```suggestion
The pattern matching function of the `sync` subcommand is similar to that of `rsync`, which allows you to exclude or include certain classes of files by rules and synchronize any set of files by combining multiple rules. Now we have the following rules available.
```

> - Patterns ending with `/` will only match directories, otherwise, they will match files, links or devices;

Reviewer: ```suggestion
- Patterns ending with `/` only matches directories. Otherwise, they match files, links or devices;
```
感觉后面半句是不是可以删去？

> - Patterns containing `*`, `?` or `[` will match as wildcards, otherwise, they will match as regular strings;

Reviewer: ```suggestion
- Patterns containing `*`, `?` or `[` follows wildcard pattern to match; otherwise, matching uses regular strings.
```
这句好难翻译，尝试了修改了一下，不知道理解对不对 :P

Author: ```suggestion
- Patterns containing `*`, `?` or `[` match as wildcards, otherwise, they match as regular strings;
```

> - `*` matches any non-empty path components until it sees a `/`;

Reviewer: ```suggestion
- `*` matches any non-empty path components (it stops at `/`).
```
借鉴了下这里 https://gist.github.com/macmladen/75817cc47f4ddf0a18f0

- `?` matches any single character except `/`;
- `[` matches a set of characters, for example `[a-z]` or `[[:alpha:]]`;

> - Backslashes can be used to escape characters in wildcard patterns, while in regular string patterns, they match literally;

Reviewer: ```suggestion
- Backslashes can be used to escape characters in wildcard patterns, while they match literally when no wildcards are present.
```

> - Match recursively using patterns as prefixes.

Reviewer: ```suggestion
- It is always matched recursively using patterns as prefixes.
```

#### Exclude Directories/Files

> Option `--exclude` can be used to exclude patterns. For example, do a full synchronization from [JuiceFS File System](#bucketC) to [Object Storage A](#bucketA), excluding hidden directories and files:

Reviewer: ```suggestion
Option `--exclude` can be used to exclude patterns. The following example shows a full synchronization from [JuiceFS File System](#bucketC) to [Object Storage A](#bucketA), excluding hidden directories and files:
```

:::note Remark

> Linux treats a directory or a file whose name starts with `.` as hidden.

Reviewer: ```suggestion
On Linux, the directory or the file with a name starting with `.` is regarded as hidden.
```

Author: ```suggestion
Linux regards a directory or a file whose name starts with `.` as hidden.
```

:::

```shell
# mount JuiceFS
sudo juicefs mount -d redis://10.10.0.8:6379/1 /mnt/jfs
# full synchronization, excluding hidden directories and files
juicefs sync --exclude '.*' /mnt/jfs/ s3://ABCDEFG:HIJKLMN@aaa.s3.us-west-1.amazonaws.com/
```

> Repeat this option with different parameters to exclude multiple patterns. For example, exclude all hidden files, `pic/` directory and `4.png` file:

Reviewer: ```suggestion
You can use this option several times with different parameters in the command to exclude multiple patterns. For example, using the following command can exclude all hidden files, `pic/` directory and `4.png` file in synchronization:
```

```shell
juicefs sync --exclude '.*' --exclude 'pic/' --exclude '4.png' /mnt/jfs/ s3://ABCDEFG:HIJKLMN@aaa.s3.us-west-1.amazonaws.com
```

#### Include Directories/Files

> Option `--include` is for including patterns you don't want to exclude. For example, synchronize only `pic/` and `4.png` and exclude all the others:

Reviewer: ```suggestion
Option `--include` can be used to include patterns you don't want to exclude. For example, only `pic/` and `4.png` are synchronized and all the others are included after executing the following command:
```

Author: ```suggestion
Option `--include` can be used to include patterns you don't want to exclude. For example, only `pic/` and `4.png` are synchronized and all the others are excluded after executing the following command:
```

```shell
juicefs sync --include 'pic/' --include '4.png' --exclude '*' /mnt/jfs/ s3://ABCDEFG:HIJKLMN@aaa.s3.us-west-1.amazonaws.com
```

:::info NOTICE

> The earlier options have higher priorities than the latter ones. Your `--include` options should appear before `--exclude`. All `--include` options such as `--include 'pic/' --include '4.png'` which appear later than `--exclude '*'` will be ignored.

Reviewer: ```suggestion
The earlier options have higher priorities than the latter ones. Thus, the `--include` options should come before `--exclude`. Otherwise, all the `--include` options such as `--include 'pic/' --include '4.png'` which appear later than `--exclude '*'` will be ignored.
```

:::

### Multi-threading and Bandwidth Throttling

> The subcommand `sync` enables 10 threads by default. You can customize thread count by `--thread` option.

Reviewer: 好像有些地方是 the subcommand `sync`，需不需要统一下格式？ 还有这句话后面的 `--thread` option?

> In addition, you can set option `--bwlimit` in the unit `Mbps` to limit the bandwidth used by the synchronization. Its default value `0` means no limitation.

Reviewer: ```suggestion
In addition, you can set option `--bwlimit` in the unit `Mbps` to limit the bandwidth used by the synchronization. The default value is `0`, meaning that bandwidth will not be constrained.
```

Author: ```suggestion
In addition, you can set option `--bwlimit` in the unit `Mbps` to limit the bandwidth used by the synchronization. The default value is `0`, meaning that bandwidth will not be limited.
```

### Directory Structure and File Permissions

The subcommand `sync` only synchronizes file objects and directories containing file objects, and skips empty directories by default. To synchronize empty directories, you can use `--dirs` option.

In addition, when synchronizing between file systems such as local, sftp and hdfs, option `--perms` can be used to synchronize file permissions from the source to the destination.

### Copy Symbolic Links

You can use `--links` option to disable symbolic link resolving when synchronizing **local directories**. That is, synchronizing only the symbolic links themselves rather than the directories or files they are pointing to. The new symbolic links created by the synchronization refer to the same paths as the original symbolic links without any conversions, no matter whether their references are reachable before or after the synchronization.

Some details need to be noticed

1. The `mtime` of a symbolic link will not be synchronized;

> 2. `--check-new` and `--perms` will be ignored on synchronizing symbolic links.

Reviewer: - 这样翻译是不是与中文的意思有些出入？
- on在这里是什么意思呢？

Author: ```suggestion
2. `--check-new` and `--perms` will be ignored when synchronizing symbolic links.
```

### Multi-machine Concurrent Synchronization

Synchronizing between two object storages is essentially pulling data from one and pushing it to the other. As shown in the figure below, the efficiency of synchronization depends on the bandwidth between the client and the cloud.

![](../images/juicefs-sync-single.png)

> When synchronizing a huge amount of data, the synchronization is often bottlenecked since the client machine runs out of bandwidth. For which case JuiceFS Sync provides a multi-machine concurrent solution, as shown in the figure below.

Reviewer: ```suggestion
When synchronizing a huge amount of data, there is often a bottleneck in the synchronization since the client machine runs out of bandwidth. For this case, JuiceFS Sync provides a multi-machine concurrent solution, as shown in the figure below.
```

![](../images/juicefs-sync-worker.png)

Manager machine executes `sync` command as the master, and defines multiple Worker machines by setting option `--worker`. JuiceFS will dynamically split the synchronization workload according to the total number of Workers and distribute to Workers for concurrent synchronization. That is, split the synchronization workload which should originally be processed on one machine into multiple parts, and dispatch them to multiple machines for concurrent processing. This increases the amount of data that can be processed per unit time, and the total bandwidth is also multiplied.

> Password-less SSH remote login from Manager to Workers should be enabled before configuring multi-machine concurrent synchronization to ensure that the client programs and the synchronization workload can be successfully distributed to Workers.

Reviewer: ```suggestion
Passwordless SSH login from Manager to Workers should be enabled before configuring multi-machine concurrent synchronization to ensure that the client programs and the synchronization workload can be successfully distributed to Workers.
```

搜了一下好像 Passwordless SSH login 更常用一些，好像这里面已经包含了与remote server/host连接的意思。如果是这样，是不是可以把remote删去呢？

:::note NOTICE
Manager distributes JuiceFS client programs to Workers. To avoid compatibility issues, please make sure the Workers use the same operating system of the same architecture as the Manager.
:::

For example, synchronize data from [Object Storage A](#bucketA) to [Object Storage B](#bucketB) concurrently with multiple machines.

```shell
juicefs sync --worker bob@192.168.1.20,tom@192.168.8.10 s3://ABCDEFG:HIJKLMN@aaa.s3.us-west-1.amazonaws.com oss://ABCDEFG:HIJKLMN@bbb.oss-cn-hangzhou.aliyuncs.com
```

The synchronization workload between the two object storages is shared by the current machine and the two Workers `bob@192.168.1.20` and `tom@192.168.8.10`.

:::tip Tips
Please set the SSH port in `.ssh/config` on the Manager machine if Workers don't listen on the default SSH port 22.
:::

## Application Scenarios

### Geo-disaster Recovery Backup

> When talking about geo-disaster recovery backup, we are talking about backing up files themselves. For example, synchronize files in [JuiceFS File System](#bucketC) to [Object Storage A](#bucketA):

Reviewer: ```suggestion
Geo-disaster recovery backup backs up files, and thus the files stored in JuiceFS should be synchronized on other object storages. For example, synchronize files in [JuiceFS File System](#bucketC) to [Object Storage A](#bucketA):
```

Author: ```suggestion
Geo-disaster recovery backup backs up files, and thus the files stored in JuiceFS should be synchronized to other object storages. For example, synchronize files in [JuiceFS File System](#bucketC) to [Object Storage A](#bucketA):
```

```shell
# mount JuiceFS
sudo juicefs mount -d redis://10.10.0.8:6379/1 /mnt/jfs
# synchronization
sudo juicefs sync /mnt/jfs/ s3://ABCDEFG:HIJKLMN@aaa.s3.us-west-1.amazonaws.com/
```

> After syncing, you can see all the files in [Object Storage A](#bucketA).

Reviewer: ```suggestion
After sync, you can see all the files in [Object Storage A](#bucketA).
```

sync也可以作名词 :)

### Build a JuiceFS Data Copy

> Unlike the file-oriented disaster recovery backup, the purpose of creating a copy of JuiceFS data is to establish a mirror with exactly the same content and structure as the JuiceFS data storage. When the object storage in use fails, you can switch to the data copy by modifying the configurations. Note that only the file data of the JuiceFS file system is replicated. The metadata stored in the metadata engine still need to be backed up.

Reviewer: ```suggestion
Unlike the file-oriented disaster recovery backup, the purpose of creating a copy of JuiceFS data is to establish a mirror with exactly the same content and structure as the JuiceFS data storage. When the object storage in use fails, you can switch to the data copy by modifying the configurations. Note that only the file data of the JuiceFS file system is replicated, and the metadata stored in the metadata engine still needs to be backed up.
```

This requires manipulating the underlying object storage directly to synchronize it with the target object storage. For example, to take the [Object Storage B](#bucketB) as the data copy of the [JuiceFS File System](#bucketC):

```shell
juicefs sync cos://ABCDEFG:HIJKLMN@ccc-125000.cos.ap-beijing.myqcloud.com oss://ABCDEFG:HIJKLMN@bbb.oss-cn-hangzhou.aliyuncs.com
```

> After syncing, the file content and hierarchy in the [Object Storage B](#bucketB) are totally the same as the [underlying object storage of JuiceFS](#bucketC).

Reviewer: ```suggestion
After sync, the file content and hierarchy in the [Object Storage B](#bucketB) are exactly the same as the [object storage of JuiceFS in use](#bucketC).
```

Author: ```suggestion
After sync, the file content and hierarchy in the [Object Storage B](#bucketB) are exactly the same as the [underlying object storage of JuiceFS](#bucketC).
```
